loadAt

navytux opened the pull request:
( This patch depends on https://github.com/zopefoundation/ZODB/pull/320 and on https://github.com/zopefoundation/ZODB/pull/322  and so  includes them it in the pull request.
   It will be rebased once, hopefully, those patches are applied to master.
   The actual change is in the patch 3, whose description is included into pull request text below for the reference )

---- 8< ----

loadAt is new optional storage interface that is intended to replace loadBefore
with more clean and uniform semantic. Compared to loadBefore, loadAt:

1) returns data=None and serial of the removal, when loaded object was found to
   be deleted. loadBefore is returning only data=None in such case. This loadAt
   property allows to fix DemoStorage data corruption when whiteouts in overlay
   part were not previously correctly taken into account.

   https://github.com/zopefoundation/ZODB/issues/318

2) for regular data records, does not require storages to return next_serial,
   in addition to (data, serial). loadBefore requirement to return both
   serial and next_serial is constraining storages unnecessarily, and,
   while for FileStorage it is free to implement, for other storages it is
   not - for example for NEO and RelStorage, finding out next_serial, after
   looking up oid@at data record, costs one more SQL query:

   https://lab.nexedi.com/nexedi/neoppod/blob/fb746e6b/neo/storage/database/mysqldb.py#L484-508
   https://lab.nexedi.com/nexedi/neoppod/blob/fb746e6b/neo/storage/database/mysqldb.py#L477-482

   https://github.com/zodb/relstorage/blob/3.1.1-1-ge7628f9/src/relstorage/storage/load.py#L259-L264
   https://github.com/zodb/relstorage/blob/3.1.1-1-ge7628f9/src/relstorage/adapters/mover.py#L177-L199

   next_serial is not only about execution overhead - it is semantically
   redundant to be there and can be removed from load return. The reason
   I say that next_serial can be removed is that in ZODB/py the only place,
   that I could find, where next_serial is used on client side is in client
   cache (e.g. in NEO client cache), and that cache can be remade to
   work without using that next_serial at all. In simple words whenever
   after

   ```
     loadAt(oid, at)  ->  (data, serial)
   ```

   query, the cache can remember data for oid in [serial, at] range.

   Next, when invalidation message from server is received, cache entries,
   that had at == client_head, are extended (at -> new_head) for oids that
   are not present in invalidation message, while for oids that are present
   in invalidation message no such extension is done. This allows to
   maintain cache in correct state, invalidate it when there is a need to
   invalidate, and not to throw away cache entries that should remain live.
   This of course requires ZODB server to include both modified and 
   just-created objects into invalidation messages

     ( https://github.com/zopefoundation/ZEO/pull/160 ,
       https://github.com/zopefoundation/ZODB/pull/319 ).

   Switching to loadAt should thus allow storages like NEO and, maybe,
   RelStorage, to do 2x less SQL queries on every object access.

   https://github.com/zopefoundation/ZODB/issues/318#issuecomment-657685745

In other words loadAt unifies return signature to always be

```
   (data, serial)
```

instead of

```
   POSKeyError                          object does not exist at all 
   None                                 object was removed
   (data, serial, next_serial)          regular data record
```

used by loadBefore.

This patch:

- introduces new interface.
- introduces ZODB.utils.loadAt helper, that uses either storage.loadAt,
  or, if the storage does not implement loadAt interface, tries to mimic
  loadAt semantic via storage.loadBefore to possible extent + emits
  corresponding warning.
- converts MVCCAdapter to use loadAt instead of loadBefore.
- changes DemoStorage to use loadAt, and this way fixes above-mentioned
  data corruption issue; adds corresponding test; converts
  DemoStorage.loadBefore to be a wrapper around DemoStorage.loadAt.
- adds loadAt implementation to FileStorage and MappingStorage.
- adapts other tests/code correspondingly.

/cc @jimfulton, @jamadden, @vpelletier, @jmuchemb, @arnaud-fontaine, @gidzit, @klawlf82, @hannosch

navytux commented:
( fixed travis wrt `ZOPE_INTERFACE_STRICT_IRO=1` )

navytux commented:
Fix CI on Windows by not forgetting to close test database before its removal. Python38 [failure](https://ci.appveyor.com/project/jimfulton/zodb/builds/34325613/job/mwi8r0rlbycnf02a) is due to _"Failed to connect to github.com port 443"_ and so should be transient.

<details>
<summary>interdiff</summary>

```diff
--- a/src/ZODB/tests/testDemoStorage.py
+++ b/src/ZODB/tests/testDemoStorage.py
@@ -357,6 +357,10 @@ def getObjAt(at):
         self.assertEqual(getObjAt(atUnlink), 2)
         self.assertRaises(ZODB.POSException.POSKeyError, getObjAt, atGC)
 
+        # end
+        db.close()
+        zdemo.close() # closes zbase and zoverlay as well
+
 
 def test_suite():
     suite = unittest.TestSuite((
```

</details>

navytux commented:
( rebased on top of master since #320 was merged )

navytux commented:
I decoupled this patch from https://github.com/zopefoundation/ZODB/pull/322 - that `ReadConflictError` -> `POSKeyError` switch requires more involvement for which I don't currently have time. On the other hand existing behaviour could be still preserved with careful programming around `loadAt` usage in `MVCCAdapter.load`. All ZODB tests pass now - without complete switch to `POSKeyError` - including `checkPackNowWhileWriting` and friends.

Once again `loadAt` is needed to fix DemoStorage data corruption (https://github.com/zopefoundation/ZODB/issues/318), and opens the door to do 2x less SQL queries on every load for NEO and RelStorage (see https://github.com/zopefoundation/ZODB/issues/318#issuecomment-657685745). I also need loadAt as the basis for further work on DemoStorage enhancements.

@jimfulton, @jamadden, @vpelletier, @jmuchemb, @arnaud-fontaine, @gidzit, @klawlf82, @hannosch,  please take a look.

Thanks beforehand,
Kirill

navytux commented:
`loadAt` is also needed for `DemoStorage.deleteObject` (https://github.com/zopefoundation/ZODB/pull/341).

navytux commented:
For the reference - see https://lab.nexedi.com/kirr/neo/commit/4fb6bd0a?expanded=1 which proides DemoStorage-analogue for ZODB/go with Load organization similar to loadAt-based DemoStorage as implemented in this pull-request.

navytux commented:
/cc @d-maurer

navytux commented:
Anyone?

d-maurer reviewed (commented):
The ZODB logic relating to historical data (including MVCC) was largely centered around `before`. You have changed this to `at` - requiring wide spread modifications. I would much prefer to keep the `before` centered approach and implement an optimized version of `loadBefore` (for backward compatibility under a new name or with a new keyword parameter) instead of introducing a new `loadAt`.

I suggest to keep `loadBefore` (without deprecation), define `loadBefore_opt` with streamlined return value and implement it by default in `BaseStorage` via `loadBefore` and change calls to `loadBefore` to calls to `loadBefore_opt` in the ZODB code base where the third return value of `loadBefore` is not necessary.

navytux commented on src/ZODB/BaseStorage.py:
```diff
@@ -267,9 +267,8 @@ def loadSerial(self, oid, serial):
-    def loadBefore(self, oid, tid):
-        """Return most recent revision of oid before tid committed."""
-        return None
+    # do not provide loadAt/loadBefore here in BaseStorage - if child forgets
+    # to override it - storage will always return "no data" instead of failing.
```
ok

navytux replied:
-> see https://github.com/navytux/ZODB/commit/d968a9a8

navytux commented:
@d-maurer, thanks for feedback.

I see about `at` vs `before` argument. Originally, when looking for a new name, I took the following into account:

- because new method differs from `loadBefore` interface behaviour, it has to come separately with different name.

- it is naturally (for me) to think of addressing ZODB database state by particular transaction ID. Given a `tid`, besides addressing a particular transaction with changes, it can also be used to specify whole database state constructed by all cumulated transaction changes from database beginning up to, and including, transaction specified by it.

  This is similar to how `commit ID` in Git addresses the state of whole source tree (would it be natural to use `commit ID` to refer to tree state as saved by parent commit?).
  And this is `at`-style addressing.

- `DB.open` already handles both `at` and `before`.
- `loadSerial(oid, serial)` implements `at`-style semantic because it loads object data as of `at=serial`, not as of `before=serial`.

From this point of view, I took the opportunity to fix `loadSerial` vs `loadBefore` inconsistency and switch everything to `at`-style addressing. Of course it was natural for me because ZODB/go uses `at`-style addressing exclusively (see e.g. https://pkg.go.dev/lab.nexedi.com/kirr/neo/go/zodb#hdr-Data_model, https://pkg.go.dev/lab.nexedi.com/kirr/neo/go/zodb#Loader, https://pkg.go.dev/lab.nexedi.com/kirr/neo/go/zodb#Xid).

Anyway, even if I prefer `at`-style, I understand that historically ZODB, at
least partly, used `before` to refer to database state, and I agree to amend my
change to follow if people prefer this way:

https://github.com/navytux/ZODB/commit/805bf36f


> I suggest to keep loadBefore (without deprecation)

I can only guess why we might not want to deprecate `loadBefore`, if new method practically provides the same or wider functionality for end users (once again, I saw `next_serial` usage only inside ZEO and NEO client storage implementations), and constraints storage implementations unnecessary to work more slowly, and to have more complex code to support not-deprecated `loadBefore`'s `next_serial` while in practice noone will likely use it.
Anyway I removed `loadBefore` deprecation warnings, but left deprecation text in the interface. Please see details in https://github.com/navytux/ZODB/commit/a640cb90.


> define `loadBefore_opt` with streamlined return value

The new name I came up with is `loadBeforeEx`. It is not `_opt` because it is
not only about optimization (not returning `next_serial`), and actually, it is
not about optimization in the first place: the new method differs from
`loadBefore` in that it returns correct serial for the case when object was
deleted. In other words here the new method provides _more_ information (this
information is needed to fix DemoStorage corruption described in
https://github.com/zopefoundation/ZODB/issues/318). The new name reflects this.


> and implement it by default in `BaseStorage` via `loadBefore`

While we can indeed imagine moving generic `loadBeforeEx` code from `ZODB.util`
into `BaseStorage`, it, unfortunately, cannot work this way. The reason it
cannot work is that storage implementations (those that implement `IStorage`)
are not obliged to inherit from `BaseStorage`. And ZEO and RelStorage provide
real examples of storage that do not inherit from `BaseStorage` at all:

https://github.com/zopefoundation/ZEO/blob/a0360167/src/ZEO/ClientStorage.py#L74-L76
https://github.com/zodb/relstorage/blob/575d61fa/src/relstorage/storage/__init__.py#L104-L109

This way we have to keep generic `loadBeforeEx` in separate utility code that
examines particular storage drivers from outside.

--------

I've made all amendments as separate steps for easier traceability, and also with the idea that if maybe later we decide to take my arguments into account and reconsider something, it is easier to revert a particular step.

Thanks again for feedback.

Kirill

d-maurer reviewed (commented):
I recognize to usefulness to have a simplified `loadBefore` variant which does not require to determine the tid for the followup modification. An `Ex` suffix may not be optimal to name this variant.
Apart from the simplification, a `loadBefore` interface weakness related to deletions has motivated the new variant. A comment in `utils` indicates that it can lead to data corruption. We could change the interface and avoid the problem: nothing would change in the absence of deletion records and the storage interface is mostly used by `ZODB` itself which we could adapt to the change.

d-maurer commented:
The PR should get an entry in `CHANGES.rst`.

d-maurer commented:
Kirill Smelkov wrote at 2021-5-4 08:02 -0700:
***@***.*** commented on this pull request.
> ...

In #174 I argued that it is not necessary to have atomic invalidation
processing for the ZEO client cache because "normal" (MVCC) loads
load only earlier object state and will not see partially invalidated
cache state.

Looking through the `loadAt` code, I noticed two places
where future object state is read:

 - the emulation of `loadBefore` via `loadBeforeEx`
   which reads all recent versions to determine the one
   following the retrieved version.

 - the `PosKeyError -> ReadConflictError` conversion
   (apparently related to concurrent pack operations)
   where future state is read to determine whether the
   object will be created in the future.

Those operations may observe partial invalidation processing.
I do not exclude that this can have bad effects.

navytux commented on src/ZODB/interfaces.py:
```diff
@@ -886,6 +889,24 @@ def prefetch(oids, tid):
         more than once.
         """
 
+class IStorageLoadBeforeEx(Interface):
+
+    def loadBeforeEx(oid, before): # -> (data, serial)
+        """Load object data as observed before given database state.
```
"Ex" stands for something like "extended". It follows (oh) approach in e.g.
Windows API where they use `<f>Ex` for extended version of `<f>`. I don't
really insist on this name. Please feel to propose a better alternative.
Something like `loadBefore2` also comes to mind.

Once again, I believe `loadAt` was good in both naming and semantic, but I
don't want to block us arguing about that.

navytux commented on src/ZODB/mvccadapter.py:
```diff
@@ -151,8 +151,19 @@ def poll_invalidations(self):
+        if data is None:
+            # raise POSKeyError if object does not exist at all
+            # TODO raise POSKeyError always and switch to raising ReadOnlyError only when
+            # actually detecting that load is being affected by simultaneous pack (see below).
+            if serial == z64:
+                # XXX second call to loadBeforeEx - it will become unneeded once we
+                # switch to raising POSKeyError.
+                _, serial_exists = loadBeforeEx(self._storage, oid, maxtid)
+                if serial_exists == z64:
+                    # object does not exist at all
+                    raise POSException.POSKeyError(oid)
+
             # object was deleted or not-yet-created.
             # raise ReadConflictError - not - POSKeyError due to backward
```
Thanks for asking.

This is indeed a bit hairy place, and I would be more happier if we could just
raise `POSKeyError` unconditionally here. We _have_ to raise `POSKeyError` when
load finds that there is no such object, because this is logically expected
behaviour when application triggers load of "not existing" object. Previously,
for "object does not exist at all", this behaviour was implemented by
`loadBefore` itself:

https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/interfaces.py#L710-L724

However raising just `POSKeyError`, when load finds that an object was deleted
as of `MVCCAdapterInstance`'s view of the database, leads to the following test
failures:

```diff
--- a/src/ZODB/mvccadapter.py
+++ b/src/ZODB/mvccadapter.py
@@ -153,6 +153,7 @@ def load(self, oid):
         assert self._start is not None
         data, serial = loadBeforeEx(self._storage, oid, self._start)
         if data is None:
+            raise POSException.POSKeyError(oid)
             # raise POSKeyError if object does not exist at all
             # TODO raise POSKeyError always and switch to raising ReadOnlyError only when
             # actually detecting that load is being affected by simultaneous pack (see below).
```

(1):
```
File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/testmvcc.py", line 413, in ZODB.tests.testmvcc
Failed example:
    r1["b"]._p_activate() # doctest: +ELLIPSIS
Expected:
    Traceback (most recent call last):
     ...
    ReadConflictError: ...
Got:
    Traceback (most recent call last):
      File "/usr/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest ZODB.tests.testmvcc[137]>", line 1, in <module>
        r1["b"]._p_activate() # doctest: +ELLIPSIS
      File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/Connection.py", line 777, in setstate
        p, serial = self._storage.load(oid)
      File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/mvccadapter.py", line 156, in load
        raise POSException.POSKeyError(oid)
    POSKeyError: 0x02
```

(2):
```
Error in test checkPackNowWhileWriting (ZODB.tests.testMappingStorage.MappingStorageTests)
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/util.py", line 265, in wrapper
    return func(*args, **kwargs)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/PackableStorage.py", line 284, in checkPackNowWhileWriting
    self._PackWhileWriting(pack_now=True)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/PackableStorage.py", line 229, in _PackWhileWriting
    t.join(60)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/MTStorage.py", line 45, in join
    self._exc_info[0], self._exc_info[1], self._exc_info[2])
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/MTStorage.py", line 37, in run
    self.runtest()
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/tests/PackableStorage.py", line 764, in runtest
    conn.root()[index].value = MinPO(j)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/Connection.py", line 777, in setstate
    p, serial = self._storage.load(oid)
  File "/home/kirr/src/wendelin/z/ZODB/src/ZODB/mvccadapter.py", line 156, in load
    raise POSException.POSKeyError(oid)
POSKeyError: 0x0a
```

The first test explicitly verifies that `ReadConflictError` is raised when
`Connection` is requested to retrieve object that exists, but was not yet
created as of connection's view of the database. It was added in 2004 in
https://github.com/zopefoundation/ZODB/commit/020daaf4cee8.

The second failure in [checkPackNowWhileWriting](https://github.com/zopefoundation/ZODB/blob/1fb097b41cae/src/ZODB/tests/PackableStorage.py#L183-L284) - that happens only sometimes - is more illustrative to "concurrent pack" problem:

- one long-running transaction is accessing the database via ZODB connection as of `@txn1` state.
- during that time, via other connection, the database is mutated with a new committed transactions (txn2) that changes object, that was already present in the database.
- then `pack(t=txn2)` is run. The pack decides to keep `obj@txn2` and deletes `obj@txn1` as instructed.
- the first transaction is still running, tries to access obj.
- load obj@txn1 finds "object exists, but was not yet created".

In other words, due to the way MVCC is implemented (snapshotted views of the
database), and in the presence of simultaneous changes, a Connection is usually
observing the database close, but not exactly, to its real head - lagging
behind it a bit. Then if pack is ran with some `t` that is inside [`conn.at`, `lastTransaction`],
for that connection it looks like the pack - concurrently to read activities of
the connection - deleted some objects under the foot. It is kind of
reader/writer conflict error.

As explained in https://github.com/zopefoundation/ZODB/commit/3a493b010876
(both commit message and changes in `mvccadapter.py`), such situations are
reported to outside via raising `ReadConflictError`.

In applications servers (e.g. Zope), transactions that raise `ConflictError`
are automatically restarted:

https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/transact.py#L25-L62

This way, if concurrent pack deletes an object under the foot of a connection,
due to `ReadConflictError` being raised, it will lead to the transaction to be
silently restarted without any error reported to user.

However, if we change to stop raising that `ReadConflictError` and instead
raise `POSKeyError` unconditionally, a real error - that `POSKeyError` - will
be reported to the user and to the log.

That's why, for backward-compatibility, the new code goes in the direction to
preserve previous behaviour of rasing `ReadConflictError` for the case when
object "exists but was not yet created".

I hope it clarifies the story of "pack and why we have to peek into the future" a bit.

Please also see the following links for more details:

- discussion starting from https://github.com/zopefoundation/ZODB/pull/320#issuecomment-661819221,
- discussion in https://github.com/zopefoundation/ZODB/pull/322.

d-maurer replied:
Satisfied with your explanation. I did not get it from the existing source code comment.

The existence of the mvcc test indicates that it may have been a design goal to raise `ReadConflictError` in the case "object only after" -- independent of pack. Maybe, the motivation was that a restarted transaction would see the object.

navytux commented:
@d-maurer, thanks for feedback.


> We could change the interface and avoid the problem: nothing would change in
> the absence of deletion records and the storage interface is mostly used by
> ZODB itself which we could adapt to the change.

I belive we should try as much as possible _not_ to change existing interface.
The reason is that it is not only users of the interface that will be affected
by an incompatible change, but also _all_ _implementers_ of the interface. In
other words even if all usage of `loadBefore` is inside ZODB (but still it is
public interface and there might be usage in other third-party places), all
storage implementations - FileStorage, ZEO, RelStorage, zc.zlibstorage,
zc.beforestorage, zc.zrs, tempstorage, ...  will _all_ _have_ to be changed
together with the interface.

Is this approach practical?

I think this case, just one more time, illustrates why they say that
"interfaces must be stable".

That's also the reason why we (imho) have to think carefully when establishing
an interface - because it is usually expensive to change an interface once it
is published and starts being used.


> In https://github.com/zopefoundation/ZEO/pull/174 I argued that it is not necessary to have atomic invalidation
> processing for the ZEO client cache because "normal" (MVCC) loads
> load only earlier object state and will not see partially invalidated
> cache state.
>
> Looking through the `loadAt` code, I noticed two places
> where future object state is read:
>
> - the emulation of `loadBefore` via `loadBeforeEx` ...
> - the `PosKeyError -> ReadConflictError` conversion ...
>
> Those operations may observe partial invalidation processing.
> I do not exclude that this can have bad effects.

Thanks for noticing this. Those operations indeed call `loadBefore` with
tid=∞ (= `maxtid`), and it can indeed probably lead to problems, as `storage.loadBefore`
is invoked with before_tid being > storage.lastTransaction leading to breakage of
the assumption that was used to justify lack of need to process invalidations atomically.

While this problem was noticed via looking at the new code added by this PR,
there are existing places in ZODB5 that use this approach:

- `DemoStorage` already has this logic to call `loadBefore(maxtid)`

  https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/DemoStorage.py#L249-L256

- `load_current` is also implemented via `loadBefore(maxtid)`:

  https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/utils.py#L371-L385

and `load_current` is used throughout ZODB in many places:

https://github.com/search?q=load_current+repo%3Azopefoundation%2FZODB&type=code

I guess the only solution to fix this is to return to process invalidations
atomically, like I was suggesting during review of https://github.com/zopefoundation/ZEO/pull/167.


Kirill

navytux commented on src/ZODB/BaseStorage.py:
```diff
@@ -269,7 +270,13 @@ def loadSerial(self, oid, serial):
+        raise NotImplementedError
+
+    def loadBeforeEx(self, oid, before):
+        """Return most recent revision of oid as of <before database state.
```
ok, let's have them both the same.

navytux replied:
-> see https://github.com/navytux/ZODB/commit/947232594c94.

navytux commented:
> The PR should get an entry in `CHANGES.rst`.

Please see https://github.com/navytux/ZODB/commit/8113c29fb1f3.

d-maurer reviewed (commented):
My favorite solution would be:
* change the `loadBefore` specification regarding the handling of object deletion records.

navytux reviewed (commented):
@d-maurer thanks for feedback.


> My favorite solution would be:

> - change the `loadBefore` specification regarding the handling of object deletion records.
This would make external storage implementations non-compliant but we could adapt the primary user of the method to ensure that this does not cause worse behavior than the current one.
> - support a simplified loadBefore (no third result component) variant as an optional optimization.


Would you please clarify this? I, frankly, don't understand how it will fully work. Even if we change ZODB code to expect that loadBefore can return both `None` or (`None`, `serial`), there are other storages - wrapper storages that I cited above - that call loadBefore and check for `None`. Do you suggest we change them all as well? And besides ZODB and storage implementations there are user code in low-level utilities that use `loadBefore` and expect return to be None as documented in the interface:

https://lab.nexedi.com/nexedi/zodbtools/blob/aa7e1966/zodbtools/zodbcommit.py
https://github.com/ZeitOnline/vivi/blob/564c27fc/core/src/zeit/edit/undo.py
https://github.com/nucypher/zerodb/blob/334be242/zerodb/permissions/ownerstorage.py
https://github.com/zopefoundation/cipher.encryptingstorage/blob/7f557a4a/src/cipher/encryptingstorage/__init__.py
...

Maybe I'm missing something, but I don't see how we could change the interface without breaking things somewhere.
That's why I suggest to keep existing interface intact and, for the new behaviour, introduce new interface.

Once again, maybe I'm missing something, so would you please explain your idea in more detail?



> > In https://github.com/zopefoundation/ZEO/pull/174 I argued that it is not necessary to have atomic invalidation
> > processing for the ZEO client cache because "normal" (MVCC) loads
> > load only earlier object state and will not see partially invalidated
> > cache state.
> > ...
> > Those operations may observe partial invalidation processing.
> > I do not exclude that this can have bad effects.

> While this problem was noticed via looking at the new code added by this PR,
> there are existing places in ZODB5 that use this approach:
> ...
> - `load_current` is also implemented via `loadBefore(maxtid)`:
> ...
> and `load_current` is used throughout ZODB in many places:
>
> https://github.com/search?q=load_current+repo%3Azopefoundation%2FZODB&type=code
>
> I guess the only solution to fix this is to return to process invalidations
> atomically, like I was suggesting during review of https://github.com/zopefoundation/ZEO/pull/167.

Dieter, would you please comment on this?


Kirill

navytux commented on CHANGES.rst:
```diff
@@ -5,6 +5,13 @@
 
+- Fix DemoStorage data corruption in the presence of whiteouts.
+  This change is implemented via introducing new ``loadBeforeEx`` interface
+  that extends and corrects ``loadBefore`` semantic.
```
Thanks. -> Please see https://github.com/navytux/ZODB/commit/c577e328.

d-maurer commented:
Kirill Smelkov wrote at 2021-5-11 10:29 -0700:
***@***.*** commented on this pull request.
***@***.*** thanks for feedback.
>> My favorite solution would be:
>
>> - change the `loadBefore` specification regarding the handling of object deletion records.
>This would make external storage implementations non-compliant but we could adapt the primary user of the method to ensure that this does not cause worse behavior than the current one.
>> - support a simplified loadBefore (no third result component) variant as an optional optimization.
>
>
>Would you please clarify this? I, frankly, don't understand how it will fully work. Even if we change ZODB code to expect that loadBefore can return both `None` or (`None`, `serial`), there are other storages - wrapper storages that I cited above - that call loadBefore and check for `None`. Do you suggest we change them all as well?

I had not expected many external (outside `ZODB`) users
of `loadBefore`.

I have not yet seen "deletion records" -- seems to be a newer concept,
potentially not yet widely used. Maybe, we can break the interface
with a version bump and an incompatibility note in the documentation?

> ...
>Maybe I'm missing something, but I don't see how we could change the interface without breaking things somewhere.

All you did was initiated by a `DemoStorage` problem. You used this
to demontrate that the `loadBefore` lacks information for
"deletion record"s. I am in favor to fix this.

>That's why I suggest to keep existing interface intact and, for the new behaviour, introduce new interface.

>Once again, maybe I'm missing something, so would you please explain your idea in more detail?

You have understood my proposal - no need to describe it again.
It will break the interface with potential consequences for
users. Let's hear what those (or others) have to say.


> ...
>> > In https://github.com/zopefoundation/ZEO/pull/174 I argued that it is not necessary to have atomic invalidation
>> > processing for the ZEO client cache because "normal" (MVCC) loads
>> > load only earlier object state and will not see partially invalidated
>> > cache state.
>> > ...
>> > Those operations may observe partial invalidation processing.
>> > I do not exclude that this can have bad effects.
>
>> While this problem was noticed via looking at the new code added by this PR,
>> there are existing places in ZODB5 that use this approach:
>> ...
>> - `load_current` is also implemented via `loadBefore(maxtid)`:
>> ...
>> and `load_current` is used throughout ZODB in many places:
>>
>> https://github.com/search?q=load_current+repo%3Azopefoundation%2FZODB&type=code
>>
>> I guess the only solution to fix this is to return to process invalidations
>> atomically, like I was suggesting during review of https://github.com/zopefoundation/ZEO/pull/167.
>
>Dieter, would you please comment on this?

I am again convinced that the loads mentioned in my original
comment do not require atomic invalidation processing:
they ignore the returned data and only look for the serial.
If the information comes form the cache, the result is the same
as if invalidation processing had been slightly postponed;
If a server load is necessary, it is automatically delayed
until invalidation processing is terminated (as invalidation
processing and server load are performed by the (mostly sequential)
IO thread).

I also do not think that `load_current` requires
atomic invalidation processing: `load_current` affects a single
oid; invalidation processing for other oids should not matter
as far as `load_current` is concerned.
Thus, as long as invalidation processing for a single oid is
atomic (ensured by the cache lock), `load_current` should be safe.

d-maurer commented:
@navytux wrote:
>     ... and that cache can be remade to
>         work without using that next_serial at all. In simple words whenever
>         after
>         ```
>           loadAt(oid, at)  ->  (data, serial)
>         ```
>         
>         
>         query, the cache can remember data for oid in [serial, at] range.
>         Next, when invalidation message from server is received, cache entries,
>         that had at == client_head, are extended (at -> new_head) for oids that
>         are not present in invalidation message,

This will not give the same cache performance as before: formerly, the cache had precise information about the currency of the object revision and for non current revisions about its validity range. Now an object revision is considered current only if it had been loaded for tid >= cache_head(at load time) (and was not invalidated since that time); however, loads for tid < cache_head can retrieve current revisions as well which is no longer recognized. For revisions loaded with tid < cache_head, the validity range is only approximated which may result in additional loads in the future.

I do not know whether the performance loss is significant in practice because most loads are likely to be for tid >= cache_head. Nevertheless, I see the potential loss as an argument against the deprecation of `loadBefore`; storages which can determine the exact validity range efficiently should be able to provide this information also in the future.

navytux commented:
@d-maurer thanks for feedback, and first of all I apologize for the delay with replying on my side.

While, probably, it was my fault not to present things clearly, let me explain
why returning only `(data, serial)` without `next_serial` should not decrease
cache performance for current - non-historical - connections:

Imagine the situation when the database is under constant change. Transactions
are committed on server and propagate via invalidations to client.
`Client.last_tid` is potentially a bit behind on-server `last_tid`. When an
on-client thread opens a `Connection`, that connection gets assigned an `at`
that denotes the Connection's snapshot of the database state. `Connection.at`,
is potentially a bit behind `Client.last_tid`, because during the time when
that client transaction is processed, client could receive some further
invalidations from the server. At any time, there might be several "current"
on-client `Connection`s that potentially all have different `Connection.at`.

In order to be efficient with putting entries into the cache, the client
maintains `ΔTail` that keeps history of all ZODB invalidations from
`Cache.head` till the "at" of its oldest "current" `Connection`. Then, when an
object is loaded for a connection with `Connection.at` < `Cache.head`, the
client looks into `ΔTail` in `(Connection.at, Cache.head]`
range to find out what is the first next serial that changed the object. If
there is an entry - that entry's serial is used as `next_serial` when putting
data into the cache after the call to load. If there is no such an entry - it
means that the data, that were loaded via `at` < `Cache.head`, is still current
even at `Cache.head` state, and so the data can be put into the cache with
`next_serial=None`.


        ΔTail.tail                            ΔTail.head

           ↓                                      ↓
    ────────────────────────────────────────────────────────────────────
                    ↑            ↑                ↑                ↑

                  conn₁        conn₂          Cache.head        on-server
                                                  =             last_tid
                                              Client.last_tid


This organization makes sure that for "current" connections - that were not
opened in historical mode - the cache will be updated in <ins>*exactly the same way*</ins>
as it would be updated if `next_serial` is returned from load.

Yes, for historical connections - that are opened via `DB.open(at/before=...)`
and are maintained in `DB.historical_pool` instead of `DB.pool` for "current"
`Connections` - the update of the cache will be less efficient compared to the
situation when `next_serial` is provided by load: the `next_serial` of the
cache entry would be set as `Connection.at` instead of real `next_serial`, thus
potentially forcing another load to happen if another historical connection is
opened with different `at`.  However, in my experience, historical connections
are seldomly used, and when used, `at` for only one particular point in time is
used to investigate a particular database state. It is thus the performance of
"current" access that matters the most.

For the reference: support for historic connections was added in late 2007 in https://github.com/zopefoundation/ZODB/commit/5f2b2d70a18a.
For the reference 2: I was talking a bit about `ΔTail` in https://github.com/zopefoundation/ZODB/issues/290#issuecomment-564988982.

----------------------------------------

> I have not yet seen "deletion records" -- seems to be a newer concept

Deletion records were added in late 2008 in https://github.com/zopefoundation/ZODB/commit/d2d83b88131b (see also http://web.archive.org/web/20120426053905/http://wiki.zope.org/ZODB/ExternalGC).

> I had not expected many external (outside `ZODB`) users of `loadBefore`.
> ...
> Maybe, we can break the interface with a version bump and an incompatibility note in the documentation?
>
> You have understood my proposal - no need to describe it again.
> It will break the interface with potential consequences for
> users. Let's hear what those (or others) have to say.

I really don't understand <ins>*why we need to break things to fix other things*</ins>,
especially when compatibility-preserving solution is possible.

Making a major version change with compatibility breakage also does not work
well in my experience. _If_ the new versions, besides breaking compatibility,
provides other benefits - that might somewhat work, but in our case the benefit
of fixed DemoStorage wrt deletion records is imho low for an average ZODB user,
while the downsides of compatibility breakage are high. This approach also
reminds me about Python2/Python3 situation where many companies are _still_
using Python2 because a) porting to Python3 has non-small cost, and b) this
porting does not usually bring any significant benefit. And all that more than
10 years after Python3 was released.

For the reference, where I work ZODB**4**
is still being used, because, even if it was not intended to be a compatibility breakage, the [switch to pickle protocol=3](https://github.com/zopefoundation/ZODB/commit/12ee41c47310) broke things in one major place for us. This, hopefully, will be addressed, but I would like to stress that changes in behaviour expose cost on users, especially changes that are intentionally backward-incompatible.

Personally I do not need, nor want, to make an experiment to see what happens if we break things.
I know beforehand that people will scream or silently get annoyed and pin ZODB to an older release.
I know the outcome of such experiment will be "this should not be the way to evolve a software".

Once again I really don't understand <ins>*why we need to break things to fix other things*</ins>,
especially when compatibility-preserving solution is possible.

Kirill

navytux commented:
> > I have not yet seen "deletion records" -- seems to be a newer concept

> Deletion records were added in late 2008 in https://github.com/zopefoundation/ZODB/commit/d2d83b88131b (see also http://web.archive.org/web/20120426053905/http://wiki.zope.org/ZODB/ExternalGC).

For the reference: deletion records can also appear without any external garbage-collection - as a result of regular undo of object creation.

navytux commented:
> For the reference: deletion records can also appear without any external garbage-collection - as a result of regular undo of object creation.

They were known and handled at least from 2002 - see https://github.com/zopefoundation/ZODB/commit/6c7d0c429b19 and:

https://github.com/zopefoundation/ZODB/blob/1fb097b41cae4ca863c8a3664414c9ec0e204393/src/ZODB/FileStorage/fsdump.py#L35-L38

d-maurer commented:
Kirill Smelkov wrote at 2021-5-21 05:31 -0700:
>... let me explain
>why returning only `(data, serial)` without `next_serial` should not decrease
>cache performance for current - non-historical - connections:
>
>Imagine the situation when the database is under constant change. Transactions
>are committed on server and propagate via invalidations to client.
>`Client.last_tid` is potentially a bit behind on-server `last_tid`. When an
>on-client thread opens a `Connection`, that connection gets assigned an `at`
>that denotes the Connection's snapshot of the database state. `Connection.at`,
>is potentially a bit behind `Client.last_tid`, because during the time when
>that client transaction is processed, client could receive some further
>invalidations from the server. At any time, there might be several "current"
>on-client `Connection`s that potentially all have different `Connection.at`.
>
>In order to be efficient with putting entries into the cache, the client
>maintains `ΔTail` that keeps history of all ZODB invalidations from
>`Cache.head` till the "at" of its oldest "current" `Connection`. Then, when an
>object is loaded for a connection with `Connection.at` < `Cache.head`, the
>client looks into `Tail` in `(Connection.at, Cache.head]`
>range to find out what is the first next serial that changed the object. If
>there is an entry - that entry's serial is used as `next_serial` when putting
>data into the cache after the call to load. If there is no such an entry - it
>means that the data, that were loaded via `at` < `Cache.head`, is still current
>even at `Cache.head` state, and so the data can be put into the cache with
>`next_serial=None`.

What are you describing above? How things currently work or how things
could be changed to be efficient without "end_tid" information?

I have looked at the ZEO [5] code and have not found a "tail":
the cache is updated in a completely straight forward way which
relies on `end_tid` to distinguish current from "historical" data.

What you are calling "client" above is actually a layered entity
where the cache is on a low (storage) layer and `Connection.at`
is maintained at a higher (`MVCCAdapter[Instance]`) layer.
Currently, the `MVCCAdapter` layer does not seem to have the API
to implement the scheme above. Likely, this should even be
implemented by the `MVCCAdapter` layer because all storages with
an associated cache would need it.

I believe (with you) that it is possible to get cache
management efficient even without "end_tid" information.
Unlike you, I fear that it requires significant (architectural) changes.
And until those changes have been done, `loadBefore` should not
get deprecated: `FileStorage` is still a major base storage and can
deliver `end_tid` information without any performance loss.


> ...
>I really don't understand <ins>*why we need to break things to fix other things*</ins>,
>especially when compatibility-preserving solution is possible.

The solution you currently have deprecates `loadBefore` in favor
of a method which does not provide validity range information.
As far as I see it, ZEO [5] currently relies on such information
for efficient cache management.
I see your argument that for some storages validity range
information is costly (and used only for special cases).
But this does not apply for e.g. `FileStorage`.
The use of its `loadBefore` should not be deprecated
until validity range information is no longer used for
cache management.


> ... in our case the benefit
>of fixed DemoStorage wrt deletion records is imho low for an average ZODB user,
Indeed, this may contribute to my reluctance to approve your PR:
it is quite large and proposes a major storage API change including the
deprecation of one of its central methods -- mainly to fix
the marginal problem above.

Of course, I, too, have proposed an (even more drastic) API change
involving potential breakage: you have found an API weakness (there
are cases when it is important to distinguish between an unknown object
and one explicitly deleted and this is not API supported) and
I thought is a good idea to fix this weakness.
When I made the proposal, I was unaware of the API callers outside the ZODB.
It would be trivial for the authors of the respective code to
adapt to the change (handle `data is None` in a straight forward way);
thus, I still feel it is a good idea - even though it is backward
incompatible.

navytux commented:
@d-maurer, thanks for feedback.

I indeed explained how things could be changed to work without `next_serial`
information, and you are right that currently there is nothing like that
implemented in ZEO: ZEO indeed currently uses `end_tid` to update its cache.

You are, again, right, that the caching layer, that I explained, is indeed
generic, and could be implemented as generic IStorage wrapper inside ZODB itself.

> I believe (with you) that it is possible to get cache
> management efficient even without "end_tid" information.

Thanks.

> Unlike you, I fear that it requires significant (architectural) changes.
> And until those changes have been done, `loadBefore` should not
> get deprecated: `FileStorage` is still a major base storage and can
> deliver `end_tid` information without any performance loss.

You are probably right here. Until we don't have that efficient caching scheme
implemented, `loadBefore` should remain in non-deprecated state.

I cannot allow myself to break things, but I have amended my change to
undeprecate `loadBefore`:

https://github.com/navytux/ZODB/commit/5ae48fe1d933

Hopefully this could help us to find some common ground.

Kirill

P.S. And `DemoStorage` is not only for demonstrations. `DemoStorage` provides generic storage overlaying, similarly to e.g. [overlayfs](https://en.wikipedia.org/wiki/OverlayFS). Besides demonstrations, such overlaying is used e.g. in tests - to create fresh ZODB instance for each test quickly - and could be used in other similar situations.

navytux commented:
@d-maurer, thanks for approval. @jamadden, do you want to provide any input on the changes?

jamadden commented:
I appreciate the reminder, and I apologize for the no-show; there are various personal and professional reasons I haven't found the time to review this PR in depth. I will do my best to look at it within the next week; feel free to move ahead without me if that doesn't happen.

I have some vague memories about some concerns I might have had:

- I'm not thrilled about the proliferation of Yet Another Load Method. In my (not very considered yet, and certainly forgetful) opinion, if this is the way forward, it should be the *only* load method (maybe that needs more work to require storages to properly support MVCC? Or the wrapper that implements it for them might need to work with storages that implement a different interface?). Yeah, it breaks backwards compatibility, but just call it 6.0 and be done with it. The overall simplification in all storage code is probably worth it. We're still supporting Python 2.7 with this, so there should be few migration blockers.
- It wasn't clear to me the impact this would have on RelStorage, since its connections are already internally consistent and have no way to load data from the future; an arbitrary "at" request is impossible to fulfill.

icemac commented:
I created some merge conflicts by merging #357, sorry.

jamadden commented:
OK, I think I'm allowed to relay this now: The company that I co-founded and which sponsored all my open source work for the last decade or so…will, umm...not be doing that…or any development work. It's been all-hands-on-deck to focus on things that could prevent that, but at this point it seems unlikely. 

I still care, deeply, about these projects. They're important to me on a personal level, and I think they are important to the wider community. But my time, at this point, is no longer going to be subsidized to work on them. I sincerely hope that changes! But in the meantime, I will have to focus on finishing the most important work I had outstanding to the community (currently I think that means gevent and greenlet), and then I have to find some way to pay the bills.

jimfulton commented:
On Wed, Nov 10, 2021 at 3:53 PM Jason Madden ***@***.***>
wrote:

> OK, I think I'm allowed to relay this now: The company that I co-founded
> and which sponsored all my open source work for the last decade or so…will,
> umm...not be doing that…or any development work. It's been
> all-hands-on-deck to focus on things that could prevent that, but at this
> point it seems unlikely.
>
I'm very sorry to hear that.  You've done a ton of great work! Thank you!!!

Jim

-- 
Jim Fulton
http://jimfulton.info

navytux commented:
@jamadden thanks for feedback. And I'm very sorry to hear that things are unfortunate on your side.
Let's hope this situation could improve for the better.

In the meantime please feel to skip what I will post below for the record.

--------

- `loadBeforeEx` intention was to replace `loadBefore` and deprecate it. It turned out we cannot deprecate loadBefore until ZEO client cache is reworked not to use `next_serial`. It is clear how to do this rework, but it was not done. For this reason loadBefore stays in undeprecated state for now:

  https://github.com/zopefoundation/ZODB/pull/323#issuecomment-842021970
  https://github.com/zopefoundation/ZODB/pull/323#issuecomment-845917355 (first part)
  https://github.com/zopefoundation/ZODB/pull/323#issuecomment-846367481 (first part)
  https://github.com/zopefoundation/ZODB/pull/323#issuecomment-855433324

- I try not to break backward compatibility. One of the motivation for loadAt/loadBeforeEx is to fix https://github.com/zopefoundation/ZODB/issues/318. I find it reasonable not to break things to fix other things, especially when backward-compatible solution is possible. The other reason is that besides ZODB and main storages (FileStorage, ZEO, NEO and RelStorage), there are many third-party places which depend on loadBefore being provided and particular semantic. It is not practical to discover and go through all those places and change them all. Also personally I find major breaking changes to be disasters, so unless such a change brings in some huge benefit, breaking backward compatibility is not worth it. In other words breaking backward-compatibility has its cost and this cost should be evaluated against expected gain. All this reasons were explained in detail in

  https://github.com/zopefoundation/ZODB/pull/323#issuecomment-833736995 (first part)
  https://github.com/zopefoundation/ZODB/pull/323#pullrequestreview-657017349 (first part)
  https://github.com/zopefoundation/ZODB/pull/323#issuecomment-845917355 (third part)
  

- the impact for RelStorage should be minimal. RelStorage already provides `loadBefore`, and `loadBeforeEx` can be implemented similarly. This implementation will have the same properties as loadBefore regarding your concern. I might be misunderstanding something but there is something like "stale handling" in RelStorage which loadBefore utilizes to detect time when SQL connection view needs to be refreshed. Once again even if RelStorage provides MVCC natively, loadBefore is already there and loadBeforeEx can be added the same way. Taking recent news into account I went ahead and sketched it myself:
  
  https://github.com/zodb/relstorage/pull/485
  
  
--------

@icemac, no problem. I resolved those conflicts.

Kirill

navytux commented:
A bit of argument to maybe reconsider `loadAt` -> `loadBeforeEx` renaming back: https://github.com/zopefoundation/ZODB/pull/341#issuecomment-973112210.

navytux commented:
( synced to master to resolve conflict on CHANGES.rst )
